src/datasources.js

```diff
@@ -30,142 +30,116 @@
-    }
-
-    function generateFixtureColumns(fixture) {
-        function generateColumns(structure) {
-            return _.map(structure, function(value, key) {
-                return [key].concat(_.map(generateColumns(value.structure), function(value) {
-                    return key + '/' + value;
-                }));
-            });
-        }
-
-        if (fixture) {
-            return _.flatten(generateColumns(fixture.structure));
-        }
-        return [];
-    }
-
-    function autocompleteChoices(fixture_uri) {
-        return generateFixtureColumns(cachedDataSources.fixture[fixture_uri]);
+    function reset() {
+        dataCache = {};
+        dataCallbacks = {};
     }
 
+    /**
+     * Asynchronously load data sources of the given type
+     *
+     * @param type - The data source type (example: "fixture").
+     * @param callback - A function to be called when the data sources
+     *      have been loaded. This function should accept two arguments:
+     *      data - {
+     *               <sourceUri1>: <data source 1 object>,
+     *               <sourceUri2>: <data source 2 object>,
+     *               ...
+     *             }
+     *      type - the data source type.
+     */
     function getDataSources(type, callback) {
-        if (!_.isEmpty(cachedDataSources[type])) {
+        if (!_.isUndefined(dataCache[type])) {
+            callback(dataCache[type], type);
+            return;
+        }
+        if (!_.isUndefined(dataCallbacks[type])) {
+            dataCallbacks[type].push(callback);
             return;
         }
 
+        function finish(data) {
+            dataCache[type] = {};
+            if (data.length === 0) {
+                dataCache[type][""] = {
+                    sourceUri: "",
+                    defaultId: "",
+                    initialQuery: "",
+                    name: "Not Found",
+                    structure: {}
+                };
+            } else {
+                _.each(data, function(item) {
+                    dataCache[type][item.sourceUri] = item;
+                });
+            }
+            _.each(dataCallbacks[type], function (callback) {
+                callback(dataCache[type], type);
+            });
+            delete dataCallbacks[type];
+        }
         var source = _.find(dataSources, function (src) {
             return src.key === type;
         });
-
+        dataCallbacks[type] = [callback];
         if (source) {
             if (_.isString(source.endpoint)) {
                 $.ajax({
                     type: 'GET',
                     url: source.endpoint,
                     dataType: 'json',
-                    success: function (data) { callback(data); },
-                    // TODO error handling
+                    success: finish,
+                    error: function (jqXHR, errorType, exc) {
+                        finish([]);
+                        window.console.log(util.formatExc(exc || errorType));
```

Comment:
Hmm, interesting idea. Problem is, this will be used in more places than just for the dropdown widget. Maybe we want to do something like the *Not Found* (when there is no data), except in this case indicate that the data could not be loaded because of a network error. Might also be nice for the user to be able to somehow retry in this case.